src/Storages/MergeTree/Compaction/MergePredicates/DistributedMergePredicate.h

```diff
@@ -0,0 +1,161 @@
+#pragma once
+
+#include <Storages/MergeTree/Compaction/MergePredicates/IMergePredicate.h>
+#include <Storages/MergeTree/MergeTreePartInfo.h>
+
+#include <base/defines.h>
+
+namespace DB
+{
+
+using PartitionIdsHint = std::unordered_set<String>;
+using CommittingBlocks = std::unordered_map<String, std::set<Int64>>;
+
+CommittingBlocks getCommittingBlocks(zkutil::ZooKeeperPtr & zookeeper, const std::string & zookeeper_path, std::optional<PartitionIdsHint> & partition_ids_hint);
+
+template<typename VirtualPartsT, typename MutationsStateT>
+class DistributedMergePredicate : public IMergePredicate
+{
+    std::expected<void, PreformattedMessage> checkCanMergePartsPreconditions(const std::string & name, const MergeTreePartInfo & info) const
```

Comment:
```suggestion
    std::expected<void, PreformattedMessage> checkCanMergePreconditionsForSinglePart(const std::string & name, const MergeTreePartInfo & info) const
```
?

Reply:
The first option is better, because it's exactly “check canMergeParts Preconditions” -- something that checks preconditions, which allows you to rely on the correctness explanation.